File: a-n-a3.php
<?php
/**
 * To apply and assess, version 3 - filter sets to Google Anaytics data,
 * incorporating support for terms as well as templates, and to exclude numbers.
 * Dauvit King <d.j.king@open.ac.uk> for ViBRANT WP3, January 2012
 */

// to see if Service Provider has a number
function match_number($search_field) {
    if (preg_match('/\d/', $search_field)) {
        return 'numeric value';
    }
    return '';
}

// to see if search terms are found in the Service Provider
// match on token characters, so dsl will match adsl for example
function match_template($search_terms, $search_field) {
    foreach ($search_terms as $token) {
        // have to test this way in case string starts at 0 = false ;-)
        $pos = strpos($search_field, $token);
        if ($pos !== false) {
            return $token;
        }
    }
    return '';
}

// to see if search tokens are found in the Service Provider
// match on exact token only, so dsl will not match adsl for example
function match_term($search_terms, $search_field) {
    foreach ($search_terms as $token) {
        if (preg_match("/\b{$token}\b/", $search_field)) {
            return $token;
        }
    }
    return '';
}

// to calculate precision, recall and f-measure
// overloaded function I know
// should be three separate functions but this is a hack ;-)
function stats($f_pos, $f_neg, $t_pos, $t_neg) {

    // precision [tp/(tp+fp)]
    $divisor = $t_pos + $f_pos;
    if ($divisor != 0) {
        $p = round(($t_pos / $divisor), 2);
    } else {
        $p = 0;
    }

    // recall [tp/(tp+fn)]
    $divisor = $t_pos + $f_neg;
    if ($divisor != 0) {
        $r = round($t_pos / $divisor, 2);
    } else {
        $r = 0;
    }

    // f-measure [(p*r)/(p+r)]
    $divisor = $p + $r;
    if ($divisor != 0) {
        $f = 2 * round(($p * $r) / $divisor, 2);
    } else {
        $f = 0;
    }

    return array($p, $r, $f);
}

// start of script proper
echo "\napply and assess started\n";

// get filter name
if ($argc == 2) {
    $filter = $argv[1];
    echo "using {$filter}\n";
} else {
    exit("\nscript exiting\nno filter file specified\n");
}

// sanity check
if (file_exists($filter . '.php')) {
    require_once($filter . '.php');
} else {
    exit("\nscript exiting\nfilter file {$filter} not found\n");
}

// suppress error messages, such as for accessing empty filter set arrays
ini_set('error_reporting', 'E_ALL & ~E_NOTICE');

// ready the stats
$tot_inc = 0;
$tp_inc = 0;
$fp_inc = 0;
$tn_inc = 0;
$fn_inc = 0;
$tot_exc = 0;
$tp_exc = 0;
$fp_exc = 0;
$tn_exc = 0;
$fn_exc = 0;
$tot_oth = 0;
$tp_oth = 0;
$fp_oth = 0;
$tn_oth = 0;
$fn_oth = 0;

// ready the files
$file_name = 'Clustering/Gold1000B';
// input
$fn_in = $file_name . '.csv';
// $fn_in = 'clustering\Gold1000B.csv';
$fin = fopen($fn_in, 'r') or exit("Unable to open input file {$fn_in}\n");

// marked output
$fn_out = $filename . $fname . '-applied.csv';
$fout = fopen($fn_out, 'w') or exit("Unable to open output file {$fn_out}\n");

// and reports
$fn_include = $file_name . '-' . $filter . '-include.txt';
$finclude= fopen($fn_include, 'w') or exit("Unable to open output file {$fn_include}\n");
$fn_exclude = $file_name . '-' . $filter . '-exclude.txt';
$fexclude = fopen($fn_exclude, 'w') or exit("Unable to open output file {$fn_exclude}\n");
$fn_other = $file_name . '-' . $filter . '-other.txt';
$fother = fopen($fn_other, 'w') or exit("Unable to open output file {$fn_other}\n");

// now classify the input one rowm at a time
while ($row = fgetcsv($fin)) {

    // set defaults 
    $pos = false;
    $row_mark = 'o';

    // test row and apply type according to filter set rules
    $token = match_template($strong_excludes, $row[0]);
    if ($token !== '') {
        $row_mark = 'e';
    } else {
        $token = match_term($strong_excludes_term, $row[0]);
        if ($token !== '') {
            $row_mark = 'e';
        } else {
            $token = match_template($includes, $row[0]);
            if ($token !== '') {
                $row_mark = 'i';
            } else {
                $token = match_term($includes_term, $row[0]);
                if ($token !== '') {
                    $row_mark = 'i';
                } else {
                    $token = match_number($row[0]);
                    if ($token !== '') {
                        $row_mark = 'e';
                    } else {
                        $token = match_template($excludes, $row[0]);
                        if ($token !== '') {
                            $row_mark = 'e';
                        } else {
                            $token = match_term($excludes_term, $row[0]);
                            if ($token !== '') {
                                $row_mark = 'e';
                            }
                        }
                    }
                }
            }
        }
    }

// mark the output

// marking matrix
// i	i	tp_inc	tn_exc	tn_oth
// i	e	fn_inc	fp_exc	tn_oth
// i	o	fn_inc	tn_exc	fp_oth
// e	i	fp_inc	fn_exc	tn_oth
// e	e	tn_inc	tp_exc	tn_oth
// e	o	tn_inc	fn_exc	fp_oth
// o	i	fp_inc	tn_exc	fn_oth
// o	e	tn_inc	fp_exc	fn_oth
// o	o	tn_inc	tn_exc	tp_oth

    $row[] = $row_mark;
    if ($row[3] == 'i') {
        $tot_inc++;
        if ($row_mark == 'i') {
            fwrite($finclude, "found {$token} in {$row[0]}\n");
            $tp_inc++;
            $tn_exc++;
            $tn_oth++;
        } else if ($row_mark == 'e') {
            fwrite($finclude, "found {$token} in {$row[0]} ** marked as exclude **\n");
            $fn_inc++;
            $fp_exc++;
            $tn_oth++;
        } else if ($row_mark == 'o') {
            fwrite($finclude, "nothing found in {$row[0]} ** marked as other **\n");
            $fn_inc++;
            $tn_exc++;
            $fp_oth++;
        }
    } else if ($row[3] =='e') {
        $tot_exc++;
        if ($row_mark == 'i') {
            fwrite($fexclude, "found {$token} in {$row[0]} ** marked as include **\n");
            $fp_inc++;
            $fn_exc++;
            $tn_oth++;
        } else if ($row_mark == 'e') {
            fwrite($fexclude, "found {$token} in {$row[0]}\n");
            $tn_inc++;
            $tp_exc++;
            $tn_oth++;
        }  else if ($row_mark == 'o') {
            fwrite($fexclude, "nothing found in {$row[0]} ** marked as other **\n");
            $tn_inc++;
            $fn_exc++;
            $fp_oth++;
        }
    } else if ($row[3] =='o') {
        $tot_oth++;
        if ($row_mark == 'i') {
            fwrite($fother, "found {$token} in {$row[0]} ** marked as include **\n");
            $fp_inc++;
            $tn_exc++;
            $fn_oth++;
        } else if ($row_mark == 'e') {
            fwrite($fother, "found {$token} in s{$row[0]} ** marked as exclude **\n");
            $tn_inc++;
            $fp_exc++;
            $fn_oth++;
        } else if ($row_mark == 'o') {
            fwrite($fother, "nothing found in {$row[0]}\n");
            $tn_inc++;
            $tn_exc++;
            $tp_oth++;
        }
    }
    // write out marked output
    fputcsv($fout, $row);
}

// prepare the stats
$sum_inc = $tp_inc + $fp_inc + $tn_inc + $fn_inc;
$sum_exc = $tp_exc + $fp_exc + $tn_exc + $fn_exc;
$sum_oth = $tp_oth + $fp_oth + $tn_oth + $fn_oth;

$stats_inc = stats($fp_inc, $fn_inc, $tp_inc, $tn_inc);
$stats_exc = stats($fp_exc, $fn_exc, $tp_exc, $tn_exc);
$stats_oth = stats($fp_oth, $fn_oth, $tp_oth, $tn_oth);

// write out the stats
$line = sprintf("\n\n%-12s %6d\n\n",'includes:', $tot_inc);
fwrite($finclude, $line);
$line = sprintf("%-12s %6d\n", 'false pos:', $fp_inc);
fwrite($finclude, $line);
$line = sprintf("%-12s %6d\n", 'false neg:', $fn_inc);
fwrite($finclude, $line);
$line = sprintf("%-12s %6d\n", 'true pos:', $tp_inc);
fwrite($finclude, $line);
$line = sprintf("%-12s %6d\n", 'true neg:', $tn_inc);
fwrite($finclude, $line);
$line = sprintf("%-12s %6d\n\n", 'sum:', $sum_inc);
fwrite($finclude, $line);
$line = sprintf("%-12s %6.2f\n", 'precision:', $stats_inc[0]);
fwrite($finclude, $line);
$line = sprintf("%-12s %6.2f\n", 'recall:', $stats_inc[1]);
fwrite($finclude, $line);
$line = sprintf("%-12s %6.2f\n", 'f-measure:', $stats_inc[2]);
fwrite($finclude, $line);

$line = sprintf("\n\n%-12s %6d\n\n",'excludes:', $tot_exc);
fwrite($fexclude, $line);
$line = sprintf("%-12s %6d\n", 'false pos:', $fp_exc);
fwrite($fexclude, $line);
$line = sprintf("%-12s %6d\n", 'false neg:', $fn_exc);
fwrite($fexclude, $line);
$line = sprintf("%-12s %6d\n", 'true pos:', $tp_exc);
fwrite($fexclude, $line);
$line = sprintf("%-12s %6d\n", 'true neg:', $tn_exc);
fwrite($fexclude, $line);
$line = sprintf("%-12s %6d\n\n", 'sum:', $sum_exc);
fwrite($fexclude, $line);
$line = sprintf("%-12s %6.2f\n", 'precision:', $stats_exc[0]);
fwrite($fexclude, $line);
$line = sprintf("%-12s %6.2f\n", 'recall:', $stats_exc[1]);
fwrite($fexclude, $line);
$line = sprintf("%-12s %6.2f\n", 'f-measure:', $stats_exc[2]);
fwrite($fexclude, $line);

$line = sprintf("\n\n%-12s %6d\n\n",'other:', $tot_oth);
fwrite($fother, $line);
$line = sprintf("%-12s %6d\n", 'false pos:', $fp_oth);
fwrite($fother, $line);
$line = sprintf("%-12s %6d\n", 'false neg:', $fn_oth);
fwrite($fother, $line);
$line = sprintf("%-12s %6d\n", 'true pos:', $tp_oth);
fwrite($fother, $line);
$line = sprintf("%-12s %6d\n", 'true neg:', $tn_oth);
fwrite($fother, $line);
$line = sprintf("%-12s %6d\n\n", 'sum:', $sum_oth);
fwrite($fother, $line);
$line = sprintf("%-12s %6.2f\n", 'precision:', $stats_oth[0]);
fwrite($fother, $line);
$line = sprintf("%-12s %6.2f\n", 'recall:', $stats_oth[1]);
fwrite($fother, $line);
$line = sprintf("%-12s %6.2f\n", 'f-measure:', $stats_oth[2]);
fwrite($fother, $line);

echo "\n              include  exclude    other\n";
$line = sprintf("%-12s %8d %8d %8d\n", 'totals:', $tot_inc, $tot_exc, $tot_oth);
echo $line;
$line = sprintf("%-12s %8d %8d %8d\n", 'false pos:', $fp_inc, $fp_exc, $fp_oth);
echo $line;
$line = sprintf("%-12s %8d %8d %8d\n", 'false neg:', $fn_inc, $fn_exc, $fn_oth);
echo $line;
$line = sprintf("%-12s %8d %8d %8d\n", 'true pos:', $tp_inc, $tp_exc, $tp_oth);
echo $line;
$line = sprintf("%-12s %8d %8d %8d\n", 'true neg:', $tn_inc, $tn_exc, $tn_oth);
echo $line;
$line = sprintf("%-12s %8d %8d %8d\n\n", 'sum:', $sum_inc, $sum_exc, $sum_oth);
echo $line;
$line = sprintf("%-12s %8.2f %8.2f %8.2f\n", 'precision:', $stats_inc[0], $stats_exc[0], $stats_oth[0]);
echo $line;
$line = sprintf("%-12s %8.2f %8.2f %8.2f\n", 'recall:', $stats_inc[1], $stats_exc[1], $stats_oth[1]);
echo $line;
$line = sprintf("%-12s %8.2f %8.2f %8.2f\n", 'f-measure:', $stats_inc[2], $stats_exc[2], $stats_oth[2]);
echo $line;

// close down the script
fclose($fin) or exit("Unable to close input file {$fn_in}\n");
fclose($fout) or exit("Unable to close output file {$fn_out}\n");
fclose($finclude) or exit("Unable to close output file {$fn_include}\n");
fclose($fexclude) or exit("Unable to close output file {$fn_exclude}\n");
fclose($fother) or exit("Unable to close output file {$fn_other}\n");

echo "\napply and assess finished\n";
?>

File: a-n-a3.py
'''Apply and assess filter sets.

David King <d.j.king@open.ac.uk> for ViBRANT <http://vbrant.eu> March 2012
'''

import csv
import re
import sys
from test_filter_sets import includes, includes_term, \
    excludes, excludes_term, strong_excludes, strong_excludes_term

def match_template(search_terms, search_field):
    '''To see if filter terms are found in the Service Provider.
       Match on term characters, so dsl will match adsl for example.'''
    for term in search_terms:
        if term in search_field:
            return term
    return False

def match_term(search_terms, search_field):
    '''To see if filter terms are found in the Service Provider.
       Match on exact term only, so dsl will not match adsl for example.'''
    for term in search_terms:
        if re.search('\\b' + term + '\\b', search_field):
            return term
    return False

def match_number(search_field):
    '''To see if a number is found in the Service Provider.
       Match on any digit anywhere in the Service Provider name.'''
    return 'number' if re.search('\d', search_field) else False

def stats(f_pos, f_neg, t_pos, t_neg):
    '''To calculate precision, recall and f-measure.
       Overloaded function I know,
       should be three separate functions but this is a hack ;-)'''

    # precision [tp/(tp+fp)]
    divisor = t_pos + f_pos
    if divisor is not 0:
        p = round((t_pos / divisor), 2)
    else:
        p = 0

    # recall [tp/(tp+fn)]
    divisor = t_pos + f_neg
    if divisor is not 0:
        r = round(t_pos / divisor, 2)
    else:
        r = 0

    # f-measure [(p*r)/(p+r)]
    divisor = p + r
    if divisor is not 0:
        f = 2 * round((p * r) / divisor, 2)
    else:
        f = 0

    return (p, r, f)

# start of script proper
print('\napply and assess started\n')

'''
# get filter name
if len(sys.argv) != 2:
    print('no name given\nscript exiting')
else:
    file_name = sys.argv[1];
'''

# ready the stats
tot_inc = tp_inc = fp_inc = tn_inc = fn_inc = 0
tot_exc = tp_exc = fp_exc = tn_exc = fn_exc = 0
tot_oth = tp_oth = fp_oth = tn_oth = fn_oth = 0

# ready the files
# set path if data not in the same directory as the script
filepath = 'Clustering/'
# core filename to expand as required below
filename = 'Gold1000B'
# and open the files
source_data = open(filepath + filename + '.csv', 'r')
filtered_data = open(filepath + filename + '-filtered.csv', 'w', newline='')
log_include = open(filepath + filename + '-include.log', 'w')
log_exclude = open(filepath + filename + '-exclude.log', 'w')
log_other = open(filepath + filename + '-other.log', 'w')

# now filter the ISPs one row at a time
reader = csv.reader(source_data)
writer = csv.writer(filtered_data)

# for each row apply type according to filter set rules
for row in reader:
    row_mark = 'o'
    matched = match_template(strong_excludes, row[0])
    if matched:
        row_mark = 'e'
    else:
        matched = match_term(strong_excludes_term, row[0])
        if matched:
            row_mark = 'e'
        else:
            matched = match_template(includes, row[0])
            if matched:
                row_mark = 'i'
            else:
                matched = match_term(includes_term, row[0])
                if matched:
                    row_mark = 'i'
                else:
                    matched = match_number(row[0])
                    if matched:
                        row_mark = 'e'
                    else:
                        matched = match_template(excludes, row[0])
                        if matched:
                            row_mark = 'e'
                        else:
                            matched = match_term(excludes_term, row[0])
                            if matched:
                                row_mark = 'e'

# mark the output
# marking matrix
# i    i    tp_inc    tn_exc    tn_oth
# i    e    fn_inc    fp_exc    tn_oth
# i    o    fn_inc    tn_exc    fp_oth
# e    i    fp_inc    fn_exc    tn_oth
# e    e    tn_inc    tp_exc    tn_oth
# e    o    tn_inc    fn_exc    fp_oth
# o    i    fp_inc    tn_exc    fn_oth
# o    e    tn_inc    fp_exc    fn_oth
# o    o    tn_inc    tn_exc    tp_oth

    row[4] = row_mark
    if row[3] == 'i':
        tot_inc += 1
        if row_mark == 'i':
            log_include.write('found {} in {}\n'.format(matched, row[0]))
            tp_inc += 1
            tn_exc += 1
            tn_oth += 1
        elif row_mark == 'e':
            log_include.write('found {} in {} ** marked as exclude **\n'.format(matched, row[0]))
            fn_inc += 1
            fp_exc += 1
            tn_oth += 1
        elif row_mark == 'o':
            log_include.write('found nothing in {} ** marked as other **\n'.format(row[0]))
            fn_inc += 1
            tn_exc += 1
            fp_oth += 1
    elif row[3] =='e':
        tot_exc += 1
        if row_mark == 'i':
            log_exclude.write('found {} in {} ** marked as include **\n'.format(matched, row[0]))
            fp_inc += 1
            fn_exc += 1
            tn_oth += 1
        elif row_mark == 'e':
            log_exclude.write('found {} in {}\n'.format(matched, row[0]))
            tn_inc += 1
            tp_exc += 1
            tn_oth += 1
        elif row_mark == 'o':
            log_exclude.write('found nothing in {} ** marked as other **\n'.format(row[0]))
            tn_inc += 1
            fn_exc += 1
            fp_oth += 1
    elif row[3] =='o':
        tot_oth += 1
        if row_mark == 'i':
            log_other.write('found {} in {} ** marked as include **\n'.format(matched, row[0]))
            fp_inc += 1
            tn_exc += 1
            fn_oth += 1
        elif row_mark == 'e':
            log_other.write('found {} in {} ** marked as exclude **\n'.format(matched, row[0]))
            tn_inc += 1
            fp_exc += 1
            fn_oth += 1
        elif row_mark == 'o':
            log_other.write('found nothing in {}\n'.format(row[0]))
            tn_inc += 1
            tn_exc += 1
            tp_oth += 1

    # write out marked output
    writer.writerow(row)

# prepare the stats
sum_inc = tp_inc + fp_inc + tn_inc + fn_inc;
sum_exc = tp_exc + fp_exc + tn_exc + fn_exc;
sum_oth = tp_oth + fp_oth + tn_oth + fn_oth;

stats_inc = stats(fp_inc, fn_inc, tp_inc, tn_inc);
stats_exc = stats(fp_exc, fn_exc, tp_exc, tn_exc);
stats_oth = stats(fp_oth, fn_oth, tp_oth, tn_oth);

# write out the stats
log_include.write('\n\n{:12s} {:6d}\n\n'.format('includes:', tot_inc))
log_include.write('{:12s} {:6d}\n'.format('false pos:', fp_inc))
log_include.write('{:12s} {:6d}\n'.format('false neg:', fn_inc))
log_include.write('{:12s} {:6d}\n'.format('true pos:', tp_inc))
log_include.write('{:12s} {:6d}\n'.format('true neg:', tn_inc))
log_include.write('{:12s} {:6d}\n\n'.format('sum:', sum_inc))
log_include.write('{:12s} {:6.2f}\n'.format('precision:', stats_inc[0]))
log_include.write('{:12s} {:6.2f}\n'.format('recall:', stats_inc[1]))
log_include.write('{:12s} {:6.2f}\n'.format('f-measure:', stats_inc[2]))

log_exclude.write('\n\n{:12s} {:6d}\n\n'.format('excludes:', tot_exc))
log_exclude.write('{:12s} {:6d}\n'.format('false pos:', fp_exc))
log_exclude.write('{:12s} {:6d}\n'.format('false neg:', fn_exc))
log_exclude.write('{:12s} {:6d}\n'.format('true pos:', tp_exc))
log_exclude.write('{:12s} {:6d}\n'.format('true neg:', tn_exc))
log_exclude.write('{:12s} {:6d}\n\n'.format('sum:', sum_exc))
log_exclude.write('{:12s} {:6.2f}\n'.format('precision:', stats_exc[0]))
log_exclude.write('{:12s} {:6.2f}\n'.format('recall:', stats_exc[1]))
log_exclude.write('{:12s} {:6.2f}\n'.format('f-measure:', stats_exc[2]))

log_other.write('\n\n{:12s} {:6d}\n\n'.format('other:', tot_oth))
log_other.write('{:12s} {:6d}\n'.format('false pos:', fp_oth))
log_other.write('{:12s} {:6d}\n'.format('false neg:', fn_oth))
log_other.write('{:12s} {:6d}\n'.format('true pos:', tp_oth))
log_other.write('{:12s} {:6d}\n'.format('true neg:', tn_oth))
log_other.write('{:12s} {:6d}\n\n'.format('sum:', sum_oth))
log_other.write('{:12s} {:6.2f}\n'.format('precision:', stats_oth[0]))
log_other.write('{:12s} {:6.2f}\n'.format('recall:', stats_oth[1]))
log_other.write('{:12s} {:6.2f}\n'.format('f-measure:', stats_oth[2]))

print('\n              include  exclude    other')
print('{:12s} {:8d} {:8d} {:8d}'.format('totals:', tot_inc, tot_exc, tot_oth))
print('{:12s} {:8d} {:8d} {:8d}'.format('false pos:', fp_inc, fp_exc, fp_oth))
print('{:12s} {:8d} {:8d} {:8d}'.format('false neg:', fn_inc, fn_exc, fn_oth))
print('{:12s} {:8d} {:8d} {:8d}'.format('true pos:', tp_inc, tp_exc, tp_oth))
print('{:12s} {:8d} {:8d} {:8d}'.format('true neg:', tn_inc, tn_exc, tn_oth))
print('{:12s} {:8d} {:8d} {:8d}\n'.format('sum:', sum_inc, sum_exc, sum_oth))
print('{:12s} {:8.2f} {:8.2f} {:8.2f}'.format('precision:', stats_inc[0], stats_exc[0], stats_oth[0]))
print('{:12s} {:8.2f} {:8.2f} {:8.2f}'.format('recall:', stats_inc[1], stats_exc[1], stats_oth[1]))
print('{:12s} {:8.2f} {:8.2f} {:8.2f}'.format('f-measure:', stats_inc[2], stats_exc[2], stats_oth[2]))

# close down the script
source_data.close()
filtered_data.close()
log_include.close()
log_exclude.close()
log_other.close()
print('\napply and assess finished\n')
